File: java/src/peachtree/aln/colourings/Colouring.java
package peachtree.aln.colourings;

import java.util.HashMap;
import java.util.Map;

import org.json.JSONArray;
import org.json.JSONObject;

import peachtree.aln.Filtering;

/**
 * Alignment colouring scheme
 * @author Jordan
 *
 */
public abstract class Colouring {
	
	
	Map<String, String> colours = new  HashMap<String, String>();
	SiteColourFilter colFilter = SiteColourFilter.all;
	Filtering filtering;
	
	public abstract String getName();
	public abstract boolean isNucleotide();
	
	
	public String getDefaultCol() {
		return "white";
	}
	
	
	
	public String getColour(String symbol, int siteNum) {
		
		
		// Major alleles only
		if (this.colFilter == SiteColourFilter.major) {
			if (!this.filtering.isMajorAllele(symbol, siteNum)) return this.getDefaultCol();
		}
		
		
		// Minor alleles only
		else if (this.colFilter == SiteColourFilter.minor) {
			if (!this.filtering.isMinorAllele(symbol, siteNum)) return this.getDefaultCol();
		}
		
		// All
		String col = this.colours.get(symbol);
		if (col == null) return getDefaultCol();
		return col;
	}
	
	
	public String toJSON() {
		JSONObject json = new JSONObject();
		for (String symbol : colours.keySet()) {
			String colour = colours.get(symbol);
			json.put(symbol, colour);
		}
		return json.toString();
		//return this.getColourMap().get(symbol);
	}
	
	@Override
	public String toString() {
		return this.getName();
	}
	
	
	/**
	 * Which sites should be coloured
	 * @param val
	 */
	public void setSiteColourFilter(SiteColourFilter val, Filtering filtering) {
		this.filtering = filtering;
		this.colFilter = val;
	}
	
	

}

File: cpp/peachtree/src/aln/colourings/Colouring.cpp
/*
 * Colouring.cpp
 *
 *  Created on: 6/07/2021
 *      Author: jdou557
 */

#include "../colourings/Colouring.h"
#include "../Filtering.h"


void Colouring::init(){
	this->colFilter = SiteColourFilter::minor;
}


void Colouring::setSiteColourFilter(SiteColourFilter val, Filtering* filtering){
	this->filtering = filtering;
	this->colFilter = val;
}


void Colouring::setSiteColourFilter(string val, Filtering* filtering){
	SiteColourFilter filterVal  = SiteColourFilter::all;
	if (val == "major") filterVal = SiteColourFilter::major;
	if (val == "minor") filterVal = SiteColourFilter::minor;
	this->setSiteColourFilter(filterVal, filtering);
}




string Colouring::getName(){
	return "null";
}


bool Colouring::isNucleotide(){
	return false;
}

string Colouring::getDefaultCol(){
	return "white";
}



/*
 * Get all values from enum
 * No elegant way of doing this AFAIK
 */
vector<string> Colouring::getSiteColourFilters(){
	vector<string> domain = {"all", "major", "minor"};
	return domain;
}
string Colouring::getDefaultSiteColourFilter(){
	return "minor";
}

string Colouring::getColour(string symbol, int siteNum){


	Filtering* f = this->filtering;

	// Major alleles only
	if (this->colFilter == SiteColourFilter::major) {
		if (!f->isMajorAllele(symbol, siteNum)) return this->getDefaultCol();
	}


	// Minor alleles only
	else if (this->colFilter == SiteColourFilter::minor) {
		if (!f->isMinorAllele(symbol, siteNum)) return this->getDefaultCol();
	}

	// All
	string col = this->colours[symbol];
	if (col == "") return this->getDefaultCol();
	return col;

}

string Colouring::toJSON(){
	json j;
	for (const auto& pair : colours) {
		string symbol = pair.first;
		string colour = colours[symbol];
		j[symbol] = colour;
	}
	return j.dump(0);

}

string Colouring::toString(){
	return this->getName();
}






